src/plonk/circuit/linear_combination.rs

```diff
@@ -716,6 +719,86 @@ impl<E: Engine> LinearCombination<E> {
     }
 }
 
+pub fn enforce_zero_naive<E: Engine, CS: ConstraintSystem<E>>(
+    cs: &mut CS,
+    terms: &[(E::Fr, Variable)], // includes value as last term
+    constant: E::Fr,
+) -> Result<(), SynthesisError> {
+    assert!(is_naive_main_gate::<E, CS>());
+    use crate::bellman::plonk::better_better_cs::cs::PlonkConstraintSystemParams;
+    assert!(CS::Params::CAN_ACCESS_NEXT_TRACE_STEP == false);
+    let chunk_size = CS::Params::STATE_WIDTH -1;
+    // c0, c1, c2 + k
+    // v0, v1, v2
+
+    // c0*v0 + c1*v1 + k = s0
+    // s0 + c2*v2 = s1
+    // s1 - sum == 0
+    let mut intermediate_sums = vec![];
+    let mut used_constant = false;
+
+    let has_value = cs.get_value(terms[0].1).is_ok();
+
+    let mut sum_terms = |chunk: &[(E::Fr, Variable)]| -> Result<(), SynthesisError> {
+        assert!(chunk.len() <= 2);
+        let mut sum = if used_constant { E::Fr::zero() } else { constant };
+        for (c, v) in chunk.iter() {
+            assert_eq!(cs.get_value(*v).is_ok(), has_value);
+            let mut tmp = cs.get_value(*v).unwrap_or(E::Fr::zero());
+            tmp.mul_assign(&c);
+            sum.add_assign(&tmp);
+        }
+
+        let allocated_sum = AllocatedNum::alloc(cs, || {
+            if has_value{
+                Ok(sum)
+            }else{
+                Err(SynthesisError::AssignmentMissing)
+            }
+        })?;
+        let mut terms = MainGateTerm::new();
+        for (coeff, var) in chunk.iter() {
+            let term = ArithmeticTerm::from_variable_and_coeff(*var, *coeff);
+            terms.add_assign(term);
+        }
+
+        if used_constant == false {
+            let term = ArithmeticTerm::constant(constant);
+            terms.add_assign(term);
+            used_constant = true;
+        }
+        let term =
+            ArithmeticTerm::from_variable_and_coeff(allocated_sum.get_variable(), E::Fr::one());
+        terms.sub_assign(term);
+
+        cs.allocate_main_gate(terms)?;
+
+        intermediate_sums.push(allocated_sum);
+
+        Ok(())
+    };
+
+    for chunk in terms.chunks_exact(chunk_size){
+        sum_terms(chunk)?;
+    }
+    // remainder term can fit into single gate
+    sum_terms(terms.chunks_exact(chunk_size).remainder())?;
+
+    assert_eq!(used_constant, true);
+    
+    if used_constant == false {
```

Comment:
While I'll be checking the layout above, this line looks very invalid (compare to false after assert to true). 

Also, you should never allocate constants (this feature is only needed for lookups), and instead add into gate.

Also, starting a sum with allocating zero variable is also not the most rational thing, you can just take first element of intermediate sums